tools/big_gc/src/phantom/gc/Main.java

```diff
@@ -136,17 +167,334 @@ private static void process(String fn) throws FileNotFoundException, IOException
 		System.out.println( String.format("Found %d objects, allocated %dK, free %dK.", 
 				objects.size(), allocatedSize/1024, freeSize/1024));
 		System.out.println("Processing.");
-		
+
 		// TODO find biggest free slot
-		
-		// process hash map and do GC
-		collectGarbage();
-		System.exit( EXIT_CODE_OK );
 	}
 
 	private static void collectGarbage() {
 		// TODO Auto-generated method stub
-		
+
+
+		// address of the very first object
+		// now it is 2147483648
+		long rootObjectAddress = OBJECT_VMEM_SHIFT;
+
+		Queue<Long> objectsToInspect = new LinkedList<>();
+		HashMap<Long, ObjectHeader> visitedObjects = new HashMap<>();
+		//HashSet<Long> visitedObjects = new HashSet<>();
+
+		objectsToInspect.add(rootObjectAddress);
+		ObjectHeader currentObject;
+		Long currentObjectAddress;
+
+
+		while (objectsToInspect.size() != 0){
+			currentObjectAddress = objectsToInspect.remove();
+			// reference isn't initialized yet
+			if(currentObjectAddress == 0) continue;
+
+
+			currentObject = objects.get(currentObjectAddress);
+			visitedObjects.put(currentObjectAddress, currentObject);
+
+			// inspect object
+			// for external objects check whether there are
+			// references to objects that hasn't been visited yet
+			// if so -> add them to the queue
+			// since internal objects has no references
```

Comment:
This is wrong, actually. Internal objects have references stored in their data areas just as struct fields. See include/vm/internal_da.h for data area structures for internal objects.